code/game/objects/items/weapons/RFD.dm

```diff
@@ -203,39 +202,50 @@ RFD Construction-Class
 		build_type =  deconstruct ? "floor" : "wall"
 		build_turf =  deconstruct ? T.baseturf : /turf/simulated/wall
 	else
-		return 0
+		return FALSE
 
 	if(!build_type)
-		working = 0
-		return 0
+		working = FALSE
+		return FALSE
+
+	if(mode == 3 && !T.density && !istype(T,/turf/simulated/floor))
+		to_chat(user, "<span class='warning'>\The [build_type] must be closed before you can deconstruct it.</span>")
+		return FALSE
 
-	if(!useResource(build_cost, user))
+	if(stored_matter < build_cost)
 		to_chat(user, SPAN_WARNING("The \'Low Ammo\' light on the device blinks yellow."))
 		flick("[icon_state]-empty", src)
-		return 0
+		return FALSE
 
-	playsound(get_turf(src), 'sound/machines/hydraulic_long.ogg', 50, 1)
+	playsound(get_turf(src), 'sound/machines/hydraulic_short.ogg', 50, 1)
 
-	working = 1
+	working = TRUE
 	user.visible_message(SPAN_NOTICE("[user] holds \the [src] towards \the [T]."), SPAN_NOTICE("You start [deconstruct ? "deconstructing" : "constructing"] \a [build_type]..."))
+	var/obj/effect/constructing_effect/rfd_effect = new(get_turf(T), src.build_delay, src.mode)
 
-	if(build_delay && !do_after(user, build_delay))
-		working = 0
-		return 0
+	if((build_delay && !do_after(user, build_delay)) || (!useResource(build_cost, user)))
+		working = FALSE
+		rfd_effect.end_animation()
+		return FALSE
 
-	working = 0
+	working = FALSE
 	if(build_delay && !can_use(user,T))
-		return 0
+		rfd_effect.end_animation()
+		return FALSE
 
 	if(build_turf)
 		T.ChangeTurf(build_turf)
-	else if(build_other)
-		new build_other(T)
 	else
 		qdel(T)
 
+	rfd_effect.end_animation()
 	playsound(get_turf(src), 'sound/effects/magnetclamp.ogg', 50, 1)
-	return 1
+	build_cost = null
+	build_delay = null
+	build_type = null
+	build_turf = null // So it resets and any fuckery is avoided
```

Comment:
why's this on the obj itself rather than in the proc?

Reply:
Actually this doesn't relate to RFD animations.

Apparently RFDs can't deconstruct airlocks if you swap between modes.

Unless you do this.